### ISSUE TITLE
[BUG] Pytorch-> Inconsistent conversion torchvision.ops.nms and onnx nonmaxsuppression

### ISSUE THREAD

**Issue by briliantnugraha** [pytorch/vision#2644]
Hi everyone, I encounter a strange issue.

I tried to convert detectron2 RoI Heads, I could convert it to ONNX successfully, but I get this error on inference:
![image](https://user-images.githubusercontent.com/9381132/92217280-16b53400-eeca-11ea-9f67-16bd4edc96d6.png)

When I check the docs of [onnx](https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonMaxSuppression), the input shapes are: boxes -> [batch, N, 4], scores -> [batch, nclass, N]
On the other hand, the format of nms in [torchvision.ops](https://pytorch.org/docs/stable/torchvision/ops.html) are:
boxes -> [N,4], scores -> [N]

I suspect that it is convesion issue in torch.onnx. Could you help me with the issue? Thanks in advance~

Code to reproduce, [[download the model](https://tinyurl.com/y3dbcqql)]:

```
import onnxruntime as rt
import numpy as np
import cv2
import os
import onnx
import torch
import torch
import sys

path_roiheads = os.path.join(path,'detectron2_roiheads.onnx') # put your frozen onnx model here
sess_roiheads = rt.InferenceSession(path_roiheads)
input_roiheads_n = [ss.name for ss in sess_roiheads.get_inputs()]
for ss in sess_roiheads.get_inputs(): print(ss.name, ss.shape, ss.type)
input_roiheads = []
for ss in sess_roiheads.get_inputs(): input_roiheads.append( np.zeros(ss.shape, dtype='float32'))
outputs_roiheads = sess_roiheads.run(None, {a:b for a,b in zip(input_roiheads_n, input_roiheads)})
```

## Expected behavior
It could run without error~

## Environment
 - PyTorch / torchvision Version (e.g., 1.0 / 0.4.0): 1.5.1/0.6.1
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch / torchvision (`conda`, `pip`, source): pip
 - Python version: 3.6.5
 - CUDA/cuDNN version: 10.0/7.6.3
 - GPU models and configuration: GTX 1080
 - More info: Numpy 1.16.4, ONNXRuntime (ORT) 1.4.0 (CPU version), ONNX 1.7.0

cc @neginraoof

**Comment by briliantnugraha** [comment]
Hi @vfdev-5 , according to this [onnx repo](https://github.com/onnx/onnx/issues/2216), it seems that the operation is not supported by default. Because I see that they use custom ops to link the NMS op with Pytorch / make constant NMS node [like this](https://github.com/onnx/onnx/issues/2216#issuecomment-667760885).

Does this mean that this op is still in the developing stage?

**Comment by vfdev-5** [comment]
Hi @briliantnugraha unfortunately, I can not help with your question. The information I know is that in torchvision, there are tests of exporting models like faster, mask rcnn and the ops : [here](https://github.com/pytorch/vision/blob/master/test/test_onnx.py#L31), [here](https://github.com/pytorch/vision/blob/master/test/test_onnx.py#L427). Could you please take a look there and check if this can help you with your issue. Thanks

**Comment by briliantnugraha** [comment]
Hi @vfdev-5 , are detectron2 and torchvision' mask-rcnn compatible?

Compatibility that I mean here is model structure, so that we could just load detectron2's model to torchvision's model.

I see that from [this discussion](https://github.com/facebookresearch/detectron2/issues/117), they should be compatible fundamentally, but my concern is that if it is different in implementation, because it will bring another issue: a conversion issue between two frameworks, which shouldn't happen in normal case.

Thanks in advance for your help~

**Comment by vfdev-5** [comment]
I think @fmassa could explain better about the compatibility between d2 and torchvision. Probably, you can have the answer by tomorrow.

In your case, you have to use a d2's model and can not use a similar model from torchvision ?

**Comment by briliantnugraha** [comment]
Great news! I will await @fmassa 's enlightnment on that topic.

Yes, we are now constrained with Detectron2. So, if there is a bridge between those two, it will be really appreciated :)

**Comment by briliantnugraha** [comment]
For reference, from my investigation between Detectron2 and Torchvision's state dicts, I find that:

- Resnet 101 layer (1 to 5) is similar
- FPN is similar
- Anchor generators are missing in the torchvision model
- RPN is similar
- RoI Heads bbox_predictor's weight and bias are different (Detectron2: 320, Torchvision: 324)
- RoI Heads mask predictor's weight and bias are different (Detectron2: 80, Torchvision: 81)

[Detectron state dict](https://github.com/pytorch/vision/files/5182374/det2_state_dict.txt)
[Torchvision state dict](https://github.com/pytorch/vision/files/5182375/torchvision_state_dict.txt)

**Comment by vfdev-5** [comment]
@briliantnugraha maybe the difference of 81 (324 = 81 * 4) vs 80 (320 = 80 * 4) is related to presence/absense of background class in MS Coco dataset...

**Comment by briliantnugraha** [comment]
Hi @vfdev-5 , yes, for mismatch size is still easy to tackle, where we could **redefine the weight and bias with our desired size**.

However, the **missing layers (i.e. anchor generators)** will be a bigger issue.

**Comment by fmassa** [comment]
@briliantnugraha the models between D2 and torchvision are different, and trying to port one to the other will be a fairly involved task (starting with the resnet structure which is different -- the stride is in a different location in the resnet blocks).

The 80 / 91 difference in the number of classes is just because we don't perform a class remapping in the COCO ids, and as the ids are not consecutive, we have 80 classes going up to id 90 or 91.

Here is an example test showing that our ONNX export for NMS works fine https://github.com/pytorch/vision/blob/a4736ea6e8ff25a1265ca5adf9e6e244d78500e8/test/test_onnx.py#L93-L102

I think it would be outside of the scope of torchvision to try to enumerate all possible implementation differences between D2. I would encourage you to first try to export torchvision model to ONNX as a test and see if it runs as expected. It might be that your version of ONNXruntime is old and doesn't support everything.

Given that most of this issue seems related to Detectron2 (and not torchvision itself), I'm closing this issue. If you manage to reproduce issues exporting torchvision models / ops to ONNX, let us know. It might be that the combination of ops in D2 makes the ONNX export fail somehow.
